vllm/compilation/fusion.py

```diff
@@ -0,0 +1,219 @@
+                             scale=scale,
+                             epsilon=1e-5)
+
+    # result
+    return at[1]
+
+
+def rms_pattern_residual_static(result: torch.Tensor, input: torch.Tensor,
+                                residual: torch.Tensor, weight: torch.Tensor,
+                                scale: torch.Tensor):
+    at = auto_functionalized(torch.ops._C.fused_add_rms_norm.default,
+                             input=input,
+                             residual=residual,
+                             weight=weight,
+                             epsilon=1e-5)
+    at1 = auto_functionalized(torch.ops._C.static_scaled_fp8_quant.default,
+                              result=result,
+                              input=at[1],
+                              scale=scale)
+
+    # result, residual
+    return at1[1], at[2]
+
+
+def rms_replacement_residual_static(result: torch.Tensor, input: torch.Tensor,
+                                    residual: torch.Tensor,
+                                    weight: torch.Tensor, scale: torch.Tensor):
+    at = auto_functionalized(
+        torch.ops._C.fused_add_rms_norm_static_fp8_quant.default,
+        result=result,
+        input=input,
+        residual=residual,
+        weight=weight,
+        scale=scale,
+        epsilon=1e-5)
+    # result, residual
+    return at[1], at[2]
+
+
+my_patterns = PatternMatcherPass()
+
+
+def empty_bf16(*args, **kwargs):
+    return torch.empty(*args, **kwargs, dtype=torch.bfloat16, device="cuda")
+
+
+def empty_fp8(*args, **kwargs):
+    return torch.empty(*args,
+                       **kwargs,
+                       dtype=torch.float8_e4m3fn,
+                       device="cuda")
+
+
+def get_patterns():
+    my_patterns = PatternMatcherPass(pass_name="fusion_pass")
+
+    inputs = [
+        empty_fp8(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(1, 5),
+        torch.empty(1, 1, device="cuda", dtype=torch.float32)
+    ]
+    register_replacement(rms_pattern_static, rms_replacement_static, inputs,
+                         fwd_only, my_patterns)
+
+    matches = []
+
+    def record_match_fn(match: Match):
+        matches.append(match)
+        return False
+
+    # with residual
+    inputs = [
+        empty_fp8(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(5, 4),
+        empty_bf16(1, 5),
+        torch.empty(1, 1, device="cuda", dtype=torch.float32)
+    ]
+    register_replacement(rms_pattern_residual_static,
+                         rms_replacement_residual_static,
+                         inputs,
+                         fwd_only,
+                         my_patterns,
+                         extra_check=record_match_fn)
+
+    return my_patterns, matches
+
+
+def process_matches(matches, graph: torch.fx.Graph):
+    for match in matches:
+        nodes = list(graph.nodes)
+        last_node_in_match = max(match.nodes, key=lambda x: nodes.index(x))
```

Comment:
wouldn't `nodes[-1]` give the same result?

Reply:
Yes, I believe so, from inspecting values in the debugger. I can look again though

Reply:
Ok, I'll buy that.  The comments in the code don't give any explicit invariants on the Match nodes so I guess we have to account for that.   I wonder if there's a more efficient way to do this though (I'm fine with figuring it out later).

Reply:
I think we could just do a single pass over the list of nodes instead of finding each one. This was just the first pass and inspired by some of the code in the pattern matcher IIRC.

Reply:
Just checked in the debugger, last node in graph is `getitem_1590`:
`match.nodes`: `[auto_functionalized_410, getitem_1587, auto_functionalized_411, getitem_1590, getitem_1588]`
Graph:
```python

     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:213 in fused_add_rms_norm, code: torch.ops._C.fused_add_rms_norm(input, residual, weight, epsilon)
    auto_functionalized_410 = torch.ops.higher_order.auto_functionalized(torch.ops._C.fused_add_rms_norm.default, input = getitem_1585, residual = getitem_1551, weight = arg479_1, epsilon = 1e-05);  getitem_1585 = getitem_1551 = arg479_1 = None
    getitem_1587: "f16[s0, 4096]" = auto_functionalized_410[1]
    getitem_1588: "f16[s0, 4096]" = auto_functionalized_410[2];  auto_functionalized_410 = None
    
     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:675 in scaled_fp8_quant, code: output = torch.empty(shape, device=input.device, dtype=out_dtype)
    empty_284: "f8e4m3fn[s0, 4096]" = torch.ops.aten.empty.memory_format([arg0_1, 4096], dtype = torch.float8_e4m3fn, device = device(type='cuda', index=0), pin_memory = False)
    
     # File: /home/luka/neuralmagic-vllm/vllm/_custom_ops.py:690 in scaled_fp8_quant, code: torch.ops._C.static_scaled_fp8_quant(output, input, scale)
    auto_functionalized_411 = torch.ops.higher_order.auto_functionalized(torch.ops._C.static_scaled_fp8_quant.default, result = empty_284, input = getitem_1587, scale = arg482_1);  empty_284 = getitem_1587 = None
    getitem_1590: "f8e4m3fn[s0, 4096]" = auto_functionalized_411[1];  auto_functionalized_411 = None
```